File: 0020_valid_parentheses/solution.py
# LeetCode 0020 - Valid Parentheses
# https://leetcode.com/problems/valid-parentheses/


class Solution:
    def isValid(self, s: str) -> bool:
        stack: list[str] = []
        pairs = {")": "(", "]": "[", "}": "{"}

        for ch in s:
            if ch in "([{":
                stack.append(ch)
            elif not stack or stack.pop() != pairs[ch]:
                return False

        return not stack

File: 0020_valid_parentheses/Solution.scala
// LeetCode 0020 - Valid Parentheses
// https://leetcode.com/problems/valid-parentheses/

object Solution {
  def isValid(s: String): Boolean = {
    val stack = scala.collection.mutable.Stack[Char]()
    val pairs = Map(')' -> '(', ']' -> '[', '}' -> '{')

    s.foreach { ch =>
      if (ch == '(' || ch == '[' || ch == '{') {
        stack.push(ch)
      } else if (stack.isEmpty || stack.pop() != pairs(ch)) {
        return false
      }
    }

    stack.isEmpty
  }
}
